### ISSUE TITLE
Help with multi-server architecture

### ISSUE THREAD

**Issue by mercmobily** [websockets/ws#1646]
Help. I realise this is not a bug, and you have any right to roll your eyes. What I can promise is that I will release everything I write about this under the GPL.
I am writing a Single Page Application would like to implement something that:

* Allows users to open different tabs, and have one collection per tab
* Allows users to send messages out to several users

I am horrified by the number of tickets open on every other high level library for socket communication. I am also horrified by the amount of configuration requirements with Docker, Nginx, AWS, etc.

I have already written the code so that each client, after connecting, receives a "tab ID" that they will keep until they reload the page (DB table `tabs`). Upon disconnection, they will reconnect and ask for updates for that tab.

In order to send a message to all tabs of a user, a record is simply added N times (once for each tab) to a database table (messagesQueue) indicating the tabId . The system periodically goes through the queue and attempts to send messages to the client. If a client hasn't received messages for 3 minutes, the tab is killed. As soon as the message is considered delivered, is deleted from the database (FIFO, obviously)

It all works really well, and I am ready to release this. But, before I do that, I need to add multi-server support.
My implementation is just so simple, that I wonder if it will work.
Short version:

* The `tabs` table ALSO has `serverIp` and `serverPort` fields
* Each server can only send stuff to clients connected to them obviously
* Each server keeps a  map of tabIds they have connections for
* Each server will only ever attempt to send messages to tabs they manage. 

So, if a message is for a tab connected to server 1.1.1.1:80, the server on 1.1.1.2:80 will not even attempt to send it.
Am I crazy, or this simple solution, which uses the database as a message queue, will "simply work" in pretty much any environment?
Am I reinventing the wheel here? Surely!

Any pearl of wisdom will be respectfully thanked!

**Comment by lpinca** [comment]
I usually do not comment on design/architecture decisions because:

1. I am not really good at it.
1. It usually depends on a specific use case, and if it works for you then it is usually ok.

That said I think the bottleneck will be the database here? Did you consider Pub/Sub?

**Comment by mercmobily** [comment]
All good, I will let you know how I go. Closing this now.
